### ISSUE TITLE
Redstone dust faces are duplicated

### ISSUE THREAD

**Issue by leduyquang753** [erich666/Mineways#93]
When exporting a Wavefront (`.obj`) model, horizontal redstone dust blocks each have two identical faces at the same position, creating a classic Z-fighting situation, for example on Blender 3.1.0 (Cycles):

![Render with duplicate faces](https://user-images.githubusercontent.com/26449060/158059537-8609b01c-6b07-4bd9-9588-b1c628bf19c2.png)

After removing one of the two faces rendering works properly:

![Render without duplicate faces](https://user-images.githubusercontent.com/26449060/158059530-a7c85f97-e921-4025-a887-72781eb8e308.png)

(Other export formats should also be checked to see whether the issue is also present.)

**Comment by erich666** [comment]
Thanks, I'll look into it tomorrow

**Comment by erich666** [comment]
Ah, I just looked at it now. What's happening is that (similar to grass and other cutouts) these redstone objects are made double-sided by drawing the top face pointing up and the bottom face pointing down. So, two faces are formed. I guess both of these are being treated as double-sided by your renderer - I wonder why this hasn't been reported earlier.

I'll look into it more tomorrow, testing with various renderers. I may need to add a "export double-sided" option...

**Comment by leduyquang753** [comment]
I just exported with the default settings, created a new Blender 3.1.0 project based on the default template, then imported the `.obj` file. The issue is present without any tweaks.
I just downloaded Blender 3.0.1 and indeed the issue is not present there, so there must have been some changes to Blender.

**Comment by erich666** [comment]
Ah, thanks! I'll give this a try.

**Comment by leduyquang753** [comment]
Here are the exported files and the Blender project: [Blender – Duplicated faces from Mineways.zip](https://github.com/erich666/Mineways/files/8333263/Blender.Duplicated.faces.from.Mineways.zip).

**Comment by erich666** [comment]
Thanks!

If I import test.obj into Blender 3.1.0, I don't see the problem:

![image](https://user-images.githubusercontent.com/803995/159780631-506a02fc-7c6b-4009-87f6-28e016c06303.png)

The cutout objects do not cast shadows properly, but the redstone is reasonable. However, I'm guessing that this might have to do with how shadows are handled and epsilons there.

Once I apply MCprep (select world's data, then "Prep Materials"), the shadows are fixed:

![image](https://user-images.githubusercontent.com/803995/159780913-90148e16-c37a-428b-ab8e-3a62340306db.png)

I _highly_ recommend applying MCprep to the model, as Blender's OBJ importer has a clear flaw (bad shadows). If you prefer doing this manually, change the Grass Material Properties, under Settings, for Shadow Mode from "Opaque" to "Alpha Hashed."

![image](https://user-images.githubusercontent.com/803995/159781569-30fa5f61-ec9c-414c-a628-06ff08682acd.png)

I _do_ see the problem you found with your test.blend file:

![image](https://user-images.githubusercontent.com/803995/159781923-4c62c56b-1686-4b01-8f6a-c377d20a6f86.png)

Looking at your material settings for any of the redstone wires, they look different than what I see. Here is what I see with your test.blend:

![image](https://user-images.githubusercontent.com/803995/159782659-66cf5b4f-43b6-4fb4-b1cc-171d8ffee245.png)

This is interesting, in that it looks like the MCprep Principled BSDF material is applied. I had chosen OptiX for Cycles. If I pick CUDA or None, it looks terrible in real-time:

![image](https://user-images.githubusercontent.com/803995/159784549-97a00094-5775-480e-ba3c-4b92ad5b851c.png)

So, it's something that is set in test.blend that I am not setting when I import your OBJ and apply MCprep to the materials. I don't know Blender well enough to know what the problem could be, unfortunately. If it's a help, here's my .blend file and textures (same as yours). 

[my_test_mcprep.zip](https://github.com/erich666/Mineways/files/8336212/my_test_mcprep.zip)

If you figure out what render or material setting is causing this z-fighting, please do let me know. I'll also let the MCprep author know about this issue, in case he has an idea. If this turns out to be a common way of setting a material, I can fix Mineways. But, given that this is the first time this issue has come up with Mineways, I'm guessing there's a workaround by setting the rendering properties another way.

Thanks for all the help. I'm glad that I can reproduce the problem with your .blend file (but not your .obj from Mineways). Now someone (not me - I'm clueless on Blender) needs to figure out what the difference is between your .blend and my, working, .blend file.

**Comment by erich666** [comment]
StandingPad on the MCprep Discord server found a fix! What you do is select the whole model, go into Edit Mode (pick it by changing Object Mode to Edit Mode in the upper left of Blender), then press M and select merge "By Distance". The problem goes away - magic. Here's an intermediate view:
![image](https://user-images.githubusercontent.com/803995/159806364-a4b9c9dc-caa3-45b5-a45d-172784ba68df.png)

and the render on your original model:

![image](https://user-images.githubusercontent.com/803995/159806515-d81aa1ba-ebe4-485b-b7af-8b326161c4db.png)

Wow. I honestly don't understand why this merge operation (documented [here](https://docs.blender.org/manual/en/2.80/modeling/meshes/editing/vertices.html)) would fix this problem - I would have thought it would *cause* z-fighting - but this seems to work for Cycles rendering.

**Comment by erich666** [comment]
He notes:

The following Python code seems to work:
import bpy
import bmesh as bm

if bpy.context.mode != 'EDIT':
    bpy.ops.object.mode_set(mode='EDIT')
    bm.select_mode = {'VERT', 'EDGE', 'FACE'} # to avoid an extra refresh 

bpy.ops.mesh.remove_doubles(1)

**Comment by StandingPadAnimations** [comment]
Code with indentation:
```py
import bpy
import bmesh as bm

if bpy.context.mode != 'EDIT':
    bpy.ops.object.mode_set(mode='EDIT')
    bm.select_mode = {'VERT', 'EDGE', 'FACE'} # to avoid an extra refresh 

bpy.ops.mesh.remove_doubles(1)
```

**Comment by StandingPadAnimations** [comment]
I've reported it as a Blender bug: https://developer.blender.org/T96744

**Comment by erich666** [comment]
Looking at the code today, I see that I had a goof: I actually did want redstone to output single sided, but had the logic reversed. If you could, @leduyquang753, please give this beta a try and let me know if it works for you: http://www.realtimerendering.com/erich/minecraft/public/mineways/downloads/Mineways_BETA1_V914.zip

**Comment by leduyquang753** [comment]
The issue is indeed resolved with the patch.

**Comment by erich666** [comment]
Great, thanks for testing it out. I think I should look more carefully at other cutouts - rails, ladders, etc. - and see if they need the same fix before releasing a new version. Anyway, I'll consider this particular case fixed.

And, thanks again for taking the time to carefully report this problem.

**Comment by erich666** [comment]
Fixed in 9.14.
